dspace-sword/src/main/java/org/dspace/sword/ItemDepositor.java

```diff
@@ -168,16 +172,17 @@ public void undoDeposit(DepositResult result) throws DSpaceSWORDException
 		try
 		{
 			SWORDContext sc = swordService.getSwordContext();
+			BundleService bundleService = ContentServiceFactory.getInstance().getBundleService();
 
 			// obtain the bitstream's owning bundles and remove the bitstream
 			// from them.  This will ensure that the bitstream is physically
 			// removed from the disk.
 			Bitstream bs = result.getBitstream();
-			Bundle[] bundles = bs.getBundles();
-			for (int i = 0; i < bundles.length; i++)
-			{
-				bundles[i].removeBitstream(bs);
-				bundles[i].update();
+			List<BundleBitstream> b2bs = bs.getBundles();
+			for (BundleBitstream b2b : b2bs) {
+				Bundle bundle = b2b.getBundle();
+				bundleService.removeBitstream(sc.getContext(), bundle, bs);
```

Comment:
I think that we will get an "ConcurrentModificationException" here. Because the "BundleBitstream" list is persisted, so when we delete an item from a list we are iterating it will alter the list.
The best way to fix this is by using an iterator:
Bitstream bs = result.getBitstream();
Iterator<BundleBitstream> b2bs = bs.getBundles().iterator();
while (b2bs.hasNext()) {
	BundleBitstream bundleBitstream = b2bs.next();
	b2bs.remove();
	Bundle bundle = bundleBitstream.getBundle();
	bundleService.removeBitstream(sc.getContext(), bundle, bs);
	bundleService.update(sc.getContext(), bundle);
}